tests/integ/test_cte.py

```diff
@@ -138,6 +138,67 @@ def test_binary(session, action):
     assert len(plan_queries["post_actions"]) == 1
 
 
+@pytest.mark.parametrize("action", binary_operations)
+def test_variable_binding_binary(session, action):
+    df1 = session.sql(
+        "select $1 as a, $2 as b from values (?, ?), (?, ?)", params=[1, "a", 2, "b"]
+    )
+    df2 = session.sql(
+        "select $1 as a, $2 as b from values (?, ?), (?, ?)", params=[1, "c", 3, "d"]
+    )
+    df3 = session.sql(
+        "select $1 as a, $2 as b from values (?, ?), (?, ?)", params=[1, "a", 2, "b"]
+    )
+
+    check_result(session, action(df1, df3), expect_cte_optimized=True)
+    check_result(session, action(df1, df2), expect_cte_optimized=False)
+
+
+def test_variable_binding_multiple(session):
+    if not session._query_compilation_stage_enabled:
+        pytest.skip(
+            "CTE query generation without the new query generation doesn't work correctly"
+        )
+
+    df1 = session.sql(
+        "select $1 as a, $2 as b from values (?, ?), (?, ?)", params=[1, "a", 2, "b"]
+    )
+    df2 = session.sql(
+        "select $1 as a, $2 as b from values (?, ?), (?, ?)", params=[1, "c", 3, "d"]
+    )
+
+    df_res = df1.union(df1).union(df2)
+    check_result(session, df_res, expect_cte_optimized=True)
+    plan_queries = df_res._plan.execution_queries
+
+    assert plan_queries[PlanQueryType.QUERIES][-1].params == [
+        1,
+        "a",
+        2,
+        "b",
+        1,
+        "c",
+        3,
+        "d",
+    ]
+
+    df_res = df2.union(df1).union(df2).union(df1)
+    check_result(session, df_res, expect_cte_optimized=True)
```

Comment:
@sfc-gh-jdu i found for this case there is only one common node detected, but by just looking at the code, there should be two common node, right? is that currently expected, if not, we can get a ticket to further investigate it

Reply:
Yeah I think so, so which one of df1 and df2 are not converted to CTE?

Reply:
df1 is converted, but not df2